src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreJob.cs

```diff
@@ -338,12 +346,40 @@ private async Task RestoreAsync(bool forceRestore, RestoreOperationSource restor
                     var pathContext = NuGetPathContext.Create(_settings);
 
                     // Get full dg spec
-                    (dgSpec, additionalMessages) = await DependencyGraphRestoreUtility.GetSolutionRestoreSpecAndAdditionalMessages(_solutionManager, cacheContext);
+                    (originalDgSpec, additionalMessages) = await DependencyGraphRestoreUtility.GetSolutionRestoreSpecAndAdditionalMessages(_solutionManager, cacheContext);
+                }
+
+                using (intervalTracker.Start(RestoreTelemetryEvent.SolutionUpToDateCheck))
+                {
+                    // Run solution based up to date check.
+                    var projectsNeedingRestore = _solutionUpToDateChecker.PerformUpToDateCheck(originalDgSpec).AsList();
+
+                    dgSpec = originalDgSpec;
+                    // Only use the optimization results if the restore is not force.
```

Comment:
spellcheck: _forced_ instead of _force_? Same few lines below with _non-force_ optimization.